### DISCUSSION TITLE
Custom AzureAD Provider - Invalid Token Type

### DISCUSSION THREAD

**Discussion by haydenisler** [nextauthjs/next-auth#2816] [Help] [answered]
Howdy everyone,

I'm using version `4.0.0-beta.2` and have set up a custom AzureAD provider to auth with my company's tenant. I'm able to successfully log in, no problems. After auth, I want to use the token generated in the authorization flow to fetch data from the `https://graph.microsoft.com/v1.0/me` endpoint. I'm using the `getToken()` function from `next-auth/jwt` and passing the raw token into the Authorization header of the request, but I'm receiving a `InvalidAuthenticationToken: Invalid Token Type.` error response from MS Graph. This fetching is happening server side in an API route. My code is the following:

`api/[...nextauth].ts`:
```typescript
import type { NextApiHandler } from 'next';
import NextAuth from 'next-auth';

const authHandler: NextApiHandler = (req, res) =>
  NextAuth(req, res, {
    providers: [
      {
        id: 'azure-ad',
        name: 'Azure Active Directory',
        type: 'oauth',
        authorization: {
          params: {
            scope:
              'offline_access openid User.Read User.Read.All User.ReadBasic.All Mail.Send',
          },
        },
        wellKnown: process.env.AZURE_AD_OPENID,
        options: {
          clientId: process.env.AZURE_AD_CLIENT_ID,
          clientSecret: process.env.AZURE_AD_CLIENT_SECRET,
        },
        profile(profile: Record<string, unknown> & { id: string }) {
          return {
            id: profile.sub,
            ...profile,
          };
        },
      },
    ],
    jwt: {
      signingKey: process.env.JWT_KEY,
    },
  });

export default authHandler;
```

`api/me/index.ts`:
```typescript
import type { NextApiRequest, NextApiResponse } from 'next';
import { getToken } from 'next-auth/jwt';

const meHandler = async (req: NextApiRequest, res: NextApiResponse) => {
  const token = await getToken({
    req,
    signingKey: process.env.JWT_KEY,
    raw: true,
  });

  const result = await fetch('https://graph.microsoft.com/v1.0/me', {
    method: 'GET',
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });

  const json = await result.json();

  res.status(200).json(json);
};

export default meHandler;
```
Any advice on how to resolve this? Thanks!

**Comment by balazsorban44** [comment]
Still working on getting an account for myself, so I can have a look at Azure AD.

**Reply by haydenisler** [reply]
Copy that, I understand the frustration with getting access to an app registration. Unfortunately, I'm not sure my company could help you in that matter, but I will ask.

Just for clarification though, the raw JWT would typically work as a bearer token? I know it's provider specific, but I'm a bit at a loss as to where to look next.

**Reply by haydenisler** [reply]
Reviewing the ms auth docs, I came across https://docs.microsoft.com/en-us/azure/active-directory/develop/id-tokens which states:

`ID tokens are intended to be understood by third-party applications. ID tokens should not be used for authorization purposes. Access tokens are used for authorization.`

My current guess is that the JWT contains the ID token and not the access token. As per https://docs.microsoft.com/en-us/azure/active-directory/develop/access-tokens:

`Be aware, however, that the tokens you receive for a Microsoft API might not always be a JWT, and that you can't always decode them.`

**Reply by haydenisler** [reply]
Update: I've gotten a successful response from the the graph endpoint. I'll post the solution in separate answer once I detail out the process.

**Comment by haydenisler** [comment, marked as answer]
I figured the issue out. I needed to persist the access token with the `jwt()` callback. Then I could use `getToken()` in my api route. If anyone else has trouble with this, I had to clear my cookies and redo the sign in process. From https://next-auth.js.org/configuration/callbacks#jwt-callback:

`The arguments user, account, profile and isNewUser are only passed the first time this callback is called on a new session, after the user signs in. In subsequent calls, only token will be available.`

My new code is as follows:

`api/[...nextauth].ts`:
```typescript
import type { NextApiHandler } from 'next';
import NextAuth from 'next-auth';

const authHandler: NextApiHandler = (req, res) =>
  NextAuth(req, res, {
    providers: [
      {
        id: 'azure-ad',
        name: 'Azure Active Directory',
        type: 'oauth',
        authorization: {
          params: {
            scope:
              'offline_access openid User.Read User.Read.All User.ReadBasic.All Mail.Send',
          },
        },
        wellKnown: process.env.AZURE_AD_OPENID,
        options: {
          clientId: process.env.AZURE_AD_CLIENT_ID,
          clientSecret: process.env.AZURE_AD_CLIENT_SECRET,
        },
        async profile(profile: Record<string, unknown> & { id: string }) {
          return {
            id: profile.sub,
            ...profile,
          };
        },
      },
    ],
    callbacks: {
      async jwt({ token, account }) {
        if (account) {
          token.accessToken = account.access_token;
        }

        return token;
      },
    },
    jwt: {
      signingKey: process.env.JWT_KEY,
    },
  });

export default authHandler;
```

`api/me/index.ts`:
```typescript
import type { NextApiRequest, NextApiResponse } from 'next';
import { getToken } from 'next-auth/jwt';

const meHandler = async (req: NextApiRequest, res: NextApiResponse) => {
  const token = await getToken({
    req,
    signingKey: process.env.JWT_KEY,
  });

  const result = await fetch('https://graph.microsoft.com/v1.0/me', {
    method: 'GET',
    headers: {
      Authorization: `Bearer ${token.accessToken}`,
    },
  });

  const json = await result.json();

  res.status(200).json(json);
};

export default meHandler;
```

This issue arose mainly out of my misunderstanding of the auth process, but I'm glad to have solved it. 👍

**Comment by balazsorban44** [comment]
We fixed Azure AD in #2818. Not yet released, but soon.
